frame/support/src/storage/bounded_vec.rs

```diff
@@ -0,0 +1,443 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2017-2021 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Traits, types and structs to support putting a bounded vector into storage, as a raw value, map
+//! or a double map.
+
+use sp_std::prelude::*;
+use sp_std::convert::TryFrom;
+use codec::{FullCodec, Encode, EncodeLike, Decode};
+use crate::{
+	traits::Get,
+	storage::{generator, StorageDecodeLength, StorageValue, StorageMap, StorageDoubleMap},
+};
+
+/// Marker trait for types `T` that can be stored in storage as `BoundedVec<T, _>`.
+pub trait BoundedVecValue: FullCodec + Clone + sp_std::fmt::Debug {}
+impl<T: FullCodec + Clone + sp_std::fmt::Debug> BoundedVecValue for T {}
+
+/// A bounded vector.
+///
+/// It has implementations for efficient append and length decoding, as with a normal `Vec<_>`, once
+/// put into storage as a raw value, map or double-map.
+///
+/// As the name suggests, the length of the queue is always bounded. All internal operations ensure
+/// this bound is respected.
+#[derive(Encode, Decode, crate::DefaultNoBound, crate::CloneNoBound, crate::DebugNoBound)]
+pub struct BoundedVec<T: BoundedVecValue, S: Get<u32>>(Vec<T>, sp_std::marker::PhantomData<S>);
+
+// NOTE: we could also implement this as:
+// impl<T: Value, S1: Get<u32>, S2: Get<u32>> PartialEq<BoundedVec<T, S2>> for BoundedVec<T, S1>
+// to allow comparison of bounded vectors with different bounds.
+impl<T: Eq + PartialEq + BoundedVecValue, S: Get<u32>> PartialEq for BoundedVec<T, S> {
```

Comment:
PartialEq shouldn't need to require Eq on T
```suggestion
impl<T: PartialEq + BoundedVecValue, S: Get<u32>> PartialEq for BoundedVec<T, S> {
```